test: add front-end tests for react code

afuetterer opened the pull request:
## Description

This PR proposes the following changes:

- add pytest-playwright as test dependency
- add a custom e2e pytest marker, disabled by default
- add django debug mode selection per env (e2e tests are run with DEBUG=True)
- add e2e tests for management navigation
- remove dedicated "webpack-build" CI job and build frontend bundle in test job
- refactor `fixtures` fixture to be reusable by other tests and fixtures
- test cases for the management UI

## How to run e2e tests

Run the e2e tests with the custom marker:
```bash
pytest -m e2e
```

Run the e2e tests with an actual visual browser:
```bash
pytest -m e2e --headed
```

Slow down the test execution:
```bash
# --slowmo: miliseconds between actions
pytest -m e2e --headed --slowmo 500
```

## Tests suggested by @MyPyDavid in #715 and related tasks

- [x] test the "Management Interface"
- [x] test navigation in #sidebar
- [x] count number of items per element type (in #list-group), as per number of items in the db (i.e. fixtures)
- [x] decide if e2e tests should run once or per Python/DB combination: yes, run only once
- [x] decide if e2e tests should run with a single browser: chromium
- [x] rebase branch
- [x] refactor tests to minimize code duplication -> parametrize
- [x] test nested display for functionality (eg. /catalogs/1/nested/)
- [x] cache browser installation in CI
- [x] rebase branch
- [x] Python 3.12 playwright (waiting for [1.39 on PyPI](https://pypi.org/project/playwright/#history), already available on GitHub)
- [x] run tests in parallel in CI with xdist: `django.db.utils.OperationalError: database table is locked`
- [x] test new item creation, fill data, save in db
- [x] test edit buttons and edit form for each element type
- [x] squash commits
- [x] add section to testing docs: https://github.com/rdmorganiser/rdmo-docs-en/pull/35
- [x] remove nested_elements
- [x] #817 is merged
- [x] rebase branch
- [x] refactor test_management_navigation() to use a for loop through model_helpers
- [x] remove properties from `ModelHelper`
- [x] squash commits

Related issue: #715

## Types of Changes
- [x] Refactoring (no functional changes, no api changes)
- [x] Build related changes

## Checklist
- [x] I have read the [contributor guide](https://github.com/rdmorganiser/rdmo/blob/master/CONTRIBUTING.md).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

afuetterer commented:
The e2e tests are slow. 
Should these tests run once in CI? Or also for Python version and Database combinations?

Should the tests run with one browser (chromium is the default), or with multiple browsers? There are a lot of options:
```console
$ playwright install --help
[...]
    Install custom browsers, supports chromium, chrome, chrome-beta, msedge, msedge-beta, msedge-dev, firefox, firefox-asan, webkit.
```

jochenklar commented:
I think one python version and one db is ok. Maybe newest python and sqlite3. Maybe edge is a good browser since nobody uses it. I use chome for development and firefox for everyday browsing.

MyPyDavid commented:
> If I click on "View catalog nested" (<- not the best name) I get the nested view, with sections, pages, questions etc. What do you mean by functionality here?

that the "View catalog nested" just works (as in functions), thought it would take too much effort to test all the elements in the nested view so we could do that later.
If you want to write this test as well, I think all the elements of a nested catalog should come out via this [CatalogNestedSerializer ](https://github.com/rdmorganiser/rdmo/blob/ee6bd484fc17437206fd8128a569dc58785da22b/rdmo/questions/serializers/v1/catalog.py#L78)

thank you very much @afuetterer, these tests look very nice already!

afuetterer commented:
> > If I click on "View catalog nested" (<- not the best name) I get the nested view, with sections, pages, questions etc. What do you mean by functionality here?
> 
> that the "View catalog nested" just works (as in functions), thought it would take too much effort to test all the elements in the nested view so we could do that later. If you want to write this test as well, I think all the elements of a nested catalog should come out via this [CatalogNestedSerializer ](https://github.com/rdmorganiser/rdmo/blob/ee6bd484fc17437206fd8128a569dc58785da22b/rdmo/questions/serializers/v1/catalog.py#L78)
> 
> thank you very much @afuetterer, these tests look very nice already!

I just wanted to know what you meant by functionality. I will let the browser click on that link and check that the new pages loads and maybe look at the first nested item. I will not test all the nested items though. That should be tested in the unit test of the serializer I guess.

afuetterer commented:
Alright, I can install msedge locally and in CI with `playwright install`, but I cannot select it with pytest-playwright, which apparently only supports these at the moment: 
```console
pytest: error: argument --browser: invalid choice: 'msedge' (choose from 'chromium', 'firefox', 'webkit')
````

I think I will stick with chromium then, which is the base of a lot of modern browsers.

afuetterer commented:
Apropos #704: of course pytest-playwright (-> playwright -> greenlet) is the only dependency that is not Python 3.12 ready. :)
I will add a Python 3.11 test run for the e2e tests for now and upgrade, once the dependency is working.

afuetterer commented:
Should playwright take a screenshot of the page in case a test fails? Just because it can?
I am unsure. These could be uploaded to GitHub as artifacts of the CI run. Overkill?

MyPyDavid commented:
I've tested it out with `pytest -m e2e --headed`, that looks pretty awesome and really fast ;)
Think that screenshots would be overkill for the moment, could use it maybe when there are multiple developers that work on features for the front-end.

afuetterer commented:
Thanks @MyPyDavid. :)

afuetterer commented:
Apropos fast, you can add slomo.

See: https://playwright.dev/python/docs/test-runners#cli-arguments

afuetterer commented:
Thank you for fixing the test. I will reword my commits soon and squash a few of them together.
This is still work in progress. But I am quiet happy with playwright so far.

I need to figure out, why editing and clicking on "save" in the edit form, does not reflect that in the database. I guess I am doing something wrong.

afuetterer commented:
Yes, not pushed yet.

```python
def test_management_edit_catalog(logged_in_admin_user: Page) -> None:
    page = logged_in_admin_user
    page.goto("/management/catalogs")
    # click edit
    page.get_by_title("Edit catalog").first.click()
    # edit
    page.get_by_label("Comment").click()
    comment = "this is a comment."
    page.get_by_label("Comment").fill(comment)
    # save
    page.get_by_role("button", name="Save").nth(1).click()

    catalog = Catalog.objects.first()
    print(catalog)
    print(catalog.comment)
    # this fails
    # assert catalog.comment == comment
```

afuetterer commented:
It writes the comment string in the comment form field and clicks save, it redirect to the catalogs page.
Maybe it is a race condition?
`Catalog.objects.first()` might be called before save is done? I will investigate. But I am on vacation for a few days.

Maybe you have more luck?

MyPyDavid commented:
ah yes, thanks for the snippet! 
sure, I can look it at, maybe I'll put a `sleep` in between ;) 
until later then!

afuetterer commented:
Playwright usually waits, you can also increase its timeout settings. I guess no need for sleeping.

afuetterer commented:
Awesome, you already found a solution? Nice.
Maybe you also know what kind of stuff people usually edit there. I just wanted to edit anything and hit save. But there is a thousand options to edit.

afuetterer commented:
If you want to play around with test generation, please try:
`python manage.py runserver` in one terminal and
`playwright codegen localhost:8000`

MyPyDavid commented:
Ive changed the target branch to 2.1.0 so that this won't be closed after closure of 2.0.1

afuetterer commented:
Yes, of course. Thank you.

MyPyDavid commented:
How does it look with this PR? I could add a test or something to the docs on Wednesday..

afuetterer commented:
I rebased to dev-2.1.0 and cleaned up my commit mess.

You could add more tests for the edit forms or the new item generation?

I will try to figure out how to reuse the authenticated state this week. The PR should be ready then. This should make the UI a least "more tested" than before. :)

Also I would like to solve the mess with the fixtures not being available in the e2e tests. This should be possible somehow.

We should also add a note on how to run the e2e tests in the rdmo docs.

afuetterer commented on rdmo/management/tests/test_frontend.py:
```diff
@@ -0,0 +1,120 @@
+    expect(page.get_by_role("heading", name="Management")).to_be_visible()
+    page.get_by_role("link", name="Catalogs").click()
+    expect(page.locator("strong").filter(has_text="Catalogs")).to_be_visible()
+    page.get_by_role("link", name="Sections").click()
```
expect page to have url ...

afuetterer commented:
@MyPyDavid: I broke the build this morning. There is a 1.39 of playwright available on GitHub, but not on PyPI yet. I guess they release it this week.

Python 3.12 should work then. No need for running e2e tests with Python 3.11.

I will fix this once the version is released.

afuetterer commented:
playwright 1.39 is available. e2e tests are possible with Python 3.12 now as well.

MyPyDavid commented:
should we make a wheels artifact from ci build maybe in this PR or in a separate one?

afuetterer commented:
Please describe what needs to be done again. You need a wheel of a branch or PR as an artifact, right?

I think it should be separate PR.

afuetterer commented:
Can you add a new issue about this to discuss the requirements?

Refs:
- https://github.com/pypa/cibuildwheel
- https://github.com/hynek/build-and-inspect-python-package

afuetterer commented:
> the test can also be finished via the front-end only:
> 
> ```python
>     page.get_by_title("Edit catalog").first.click()
>     comment_locator = page.get_by_label("Comment")
>     expect(comment_locator).to_have_text(comment)
>     # Catalog
>     catalog = Catalog.objects.first()
>     # now it passes
>     assert catalog.comment == comment
> ```
> 
> but now also the db `assert` works.. think that the `expect` executes the save to the db and before that it is lazily evualated?

I cannot reproduce this. Can you show me the complete test function, that actually saves the comment in the db?

MyPyDavid commented:
Yes, here it is, this testcase passes for me:
```py

def test_management_edit_catalog(logged_in_admin_user: Page) -> None:
    page = logged_in_admin_user
    page.goto("/management/catalogs")
    # click edit
    page.get_by_title("Edit catalog").first.click()
    # edit
    page.get_by_label("Comment").click()
    comment = "this is a comment."
    page.get_by_label("Comment").fill(comment)
    # save
    page.get_by_role("button", name="Save").nth(1).click()
    # click on edit catalog again
    page.get_by_title("Edit catalog").first.click()
    # check the updated comment
    comment_locator = page.get_by_label("Comment")
    expect(comment_locator).to_have_text(comment)
    # compare the updated comment with catalog from db
    catalog = Catalog.objects.first()
    assert catalog.comment == comment

```

MyPyDavid commented on rdmo/management/tests/test_frontend.py:
```diff
@@ -0,0 +1,186 @@
+    page.get_by_label(helper.form_field).fill(value)
+    # save
+    page.get_by_role("button", name="Create").nth(1).click()
+
```
if you add this here, it should pass:
```py
    items_in_ui = page.locator(".list-group > .list-group-item")
    expect(items_in_ui).to_have_count(num_objects_at_start + 1)
```

MyPyDavid commented on rdmo/management/tests/test_frontend.py:
```diff
@@ -0,0 +1,186 @@
+    page.get_by_role("button", name="Create").nth(1).click()
+
+    num_objects_after_save = helper.model.objects.count()
+    assert num_objects_after_save > num_objects_at_start
```
could also assert like this:
```py

    assert num_objects_after_save - num_objects_at_start == 1
```

afuetterer commented:
Thanks @MyPyDavid for the additions. I will look at the "reusing auth issue" again.

I would like to fix my messed up commit messages, so I will force-push soon, if thats alright.

Do you feel confident that these e2e tests do something with the frontend so far? I guess this is a great step in the right direction, e.g. the playwright setup is in place. More tests can alwasy be added. But this is so much better than no tests for the frontend at all.

afuetterer commented:
> Can you add a new issue about this to discuss the requirements?
> 
> Refs:
> 
>     * https://github.com/pypa/cibuildwheel
> 
>     * https://github.com/hynek/build-and-inspect-python-package

Can you create the new issue?

MyPyDavid commented:
yes, I think that the tests are good so far. Playwright looks like a nice framework.
The locators in the test cases can be found interactively with:
```
playwright codegen --viewport-size=800,600 playwright.dev
```
The "reusing auth issue" is an optimization I guess, it can also be added later in combination with extending the tests for multiple users.

thank you @afuetterer !

MyPyDavid commented:
> > Can you add a new issue about this to discuss the requirements?
> > Refs:
> > ```
> > * https://github.com/pypa/cibuildwheel
> > 
> > * https://github.com/hynek/build-and-inspect-python-package
> > ```
> 
> Can you formulate the new issue with the requirements?

I've posted this issue https://github.com/rdmorganiser/rdmo/issues/802 about the wheels

afuetterer commented:
Thank you @MyPyDavid. I totally agree. Lets wrap this up, then. I will add a new issue about reusing authentication state,

Will squash my commits later today.

MyPyDavid commented on rdmo/management/tests/test_frontend.py:
```diff
@@ -0,0 +1,210 @@
+    expect(page).to_have_url(re.compile(r".*/views/"))
+
+
+nested_elements = (
```
this `nested_elements` is not used anymore, it can be removed now.
The propery is included in the `model_helpers` , `has_nested` kwarg.

afuetterer replied:
I see, thank you. Will remove it then.

Do you think the ModelHelper is a good idea for this? I just took a dataclass to have some information at hand and not to have so much duplication in the test functions.

afuetterer replied:
Done.

afuetterer replied:
Alright, I can add these methods to the helper.

afuetterer replied:
I added the properties. Where do you need verbose name and plural?

afuetterer commented:
What it the timing for 2.0.2? Should this PR be a part of 2.0.2 as well?

jochenklar commented:
No this can wait, idealy 2.0.2 will be published tomorrow.

MyPyDavid commented on rdmo/management/tests/test_frontend.py:
```diff
@@ -31,19 +31,32 @@
+        return f"{self.model._meta.model_name}s"
+
     @property
     def url_name(self) -> str:
```
the `verbose_name` replaces the `url_name`

MyPyDavid commented on rdmo/management/tests/test_frontend.py:
```diff
@@ -0,0 +1,213 @@
+    yield page
+
+
+def test_management_navigation(logged_in_admin_user: Page) -> None:
```
thought to make a for-loop of `model_helper` inside this test function, that could use the `verbose_name_plural` as value for eg. `name="Question sets"` , it might need a `.capitalize` to get `Questions sets` instead of `Question Sets`

afuetterer replied:
Is this wrong?

https://github.com/rdmorganiser/rdmo/blob/a54867e9604bf7ed604ff45ce219bbc33f5ff21e/rdmo/questions/models/questionset.py#L188

MyPyDavid replied:
yes that looks like a bug 😉 
In the models there is actually also a `verbose_name` and `_plural` property that include the translation as well..
https://github.com/rdmorganiser/rdmo/blob/a54867e9604bf7ed604ff45ce219bbc33f5ff21e/rdmo/questions/models/questionset.py#L205
then those could be used directly instead of the properties on the `ModelHelper` 🙈

afuetterer replied:
Alright. The Modelhelper itself is still useful, right?
I will remove the properties and call the model.verbose_name stuff directly.

Also I added a fix PR for the questionset verbose name in #817

afuetterer replied:
What do you think @MyPyDavid? What info should remain in the modelhelper class?

MyPyDavid replied:
think its fine like that, the properties on the modelhelper as "shortcuts" make it more readable anyways

MyPyDavid commented on rdmo/management/tests/test_frontend.py:
```diff
@@ -0,0 +1,213 @@
+    page = logged_in_admin_user
+    expect(page.get_by_role("heading", name="Management")).to_be_visible()
+    page.get_by_role("link", name="Catalogs").click()
+    expect(page).to_have_url(re.compile(r".*/catalogs/"))
```
do you want to make a `for` loop over the `model_helpers` from here on in this test method?

afuetterer replied:
Yes. You proposed this before. I think it broke because of the wrong question set verbose name plural, which is now fixed. I will push a for loop for this method before Thursday.
Maybe we can merge this PR then. And after that we can check if those JS dependency updates break the frontend. Hopefully this UI test will catch something.

MyPyDavid replied:
yes exactly, that would be great!

afuetterer commented:
Done with this for now. Tests are passing.

MyPyDavid reviewed (approved):
that is awesome! thanks so much XD

afuetterer commented:
> that is awesome! thanks so much XD

You are welcome.

afuetterer commented:
@jochenklar: This is ready for review. I think it will help e.g. finding breaking updates in newer javascript library versions.

jochenklar commented on rdmo/questions/models/questionset.py:
```diff
@@ -185,7 +185,7 @@ class QuestionSet(Model, TranslationMixin):
         ordering = ('uri', )
         verbose_name = _('Question set')
-        verbose_name_plural = _('Question set')
+        verbose_name_plural = _('Question sets')
```
Don't we have another PR for this, I don't mind merging it here, but then we need to drop the other PR.

afuetterer replied:
Yes, that is right. I will rebase.

jochenklar reviewed (approved):
Awesome, thanks this looks very good!
